rezasm-source/rezasm-core/src/parser/line.rs

```diff
@@ -34,35 +37,61 @@ impl Line {
-        for arg in args {
-            if looks_like_string_immediate(&arg) {
-                args_out.push(Token::StringImmediate(arg));
-            } else if looks_like_dereference(&arg) {
-                args_out.push(get_dereference(&arg)?);
-            } else if looks_like_character_immediate(&arg) {
-                args_out.push(get_character_immediate(&arg)?);
-            } else if looks_like_numerical_immediate(&arg) {
-                args_out.push(get_numerical_immediate(&arg)?);
-            } else if is_register(&arg) {
-                args_out.push(get_register(&arg)?);
-            } else if looks_like_label_reference(&arg) {
-                args_out.push(Token::LabelReference(arg));
+        for arg in &args {
+            if looks_like_string_immediate(arg) {
+                args_out.push(Token::StringImmediate(arg.to_string()));
+            } else if looks_like_dereference(arg) {
+                args_out.push(get_dereference(arg)?);
+            } else if looks_like_character_immediate(arg) {
+                args_out.push(get_character_immediate(arg)?);
+            } else if looks_like_numerical_immediate(arg) {
+                args_out.push(get_numerical_immediate(arg)?);
+            } else if is_register(arg) {
+                args_out.push(get_register(arg)?);
+            } else if looks_like_label_reference(arg) {
+                args_out.push(Token::LabelReference(arg.to_string()));
             } else {
-                return Err(ParserError::UnknownTokenError(arg).into());
+                return Err(ParserError::UnknownTokenError(arg.to_string()).into());
             }
         }
 
         let mut arguments: Vec<ArgumentType> = Vec::new();
-        for argument in args_out {
-            arguments.push(match argument.get_target(word_size) {
-                Ok(arg) => arg,
-                Err(error) => return Err(error),
-            });
-        }
-
-        let matching_instruction = get_instruction(instruction, &arguments)?;
+        let instruction_retrieved =
+            match instruction_registry::get_instruction(instruction, args_out.len()) {
+                Ok(instruction) => instruction,
+                Err(e) => return Err(e),
+            };
 
-        Ok(Line::Instruction(matching_instruction, arguments))
+        for (index, (argument, type_of)) in args_out
+            .iter()
+            .zip(instruction_retrieved.get_types())
+            .enumerate()
+        {
+            if type_of == &TypeId::of::<&mut InputTarget>() {
+                arguments.push(argument.get_input_target(word_size)?);
+            } else if type_of == &TypeId::of::<&mut InputOutputTarget>() {
+                arguments.push(match argument.get_input_output_target(word_size) {
+                    Ok(s) => s,
+                    Err(e) => {
+                        return match e {
+                            ParserError::InternalError => Err(ParserError::InvalidArgumentsError(
```

Comment:
if this `InvalidArgumentsError` percolates to an end user, it will be giving them incorrect information about why their program isn't working. In reality, this should only ever happen in the case of a bug in our parsing of correct code or in the case of some unexpected behavior.

Refer to other discussion about potentially reworking `ParserError::InternalError` into its own enum type.